src/database/error/UniqueViolation.js

```diff
@@ -0,0 +1,5 @@
+import DatabaseError from './DatabaseError.js';
+
+export default class UniqueViolation extends DatabaseError {}
+
+DatabaseError.define('23505', UniqueViolation);
```

Comment:
Error codes are found [here](https://www.postgresql.org/docs/13/errcodes-appendix.html). As we run into them we can create classes for them and then improve error handling in our code.

When data is validated by the database, I usually just rely on the database to do that validation. By properly checking the codes (and other properties if necessary), we can determine exactly why the data was invalid and give a better error message to the user. 

Honestly, I'm not 100% sure on the best practice here, some people might say it's better to avoid hitting the database with errors, but if you try to do that you have a bunch of problems:
1. You have to duplicate the check at the code level, and it's easy to implement the duplicate wrong
2. If you change the constraint at the database level, you have to the duplicate again
3. You have to load extra data from the database anyway in order to perform that check

The other side of the argument:
1. Validation is business logic, which should not be in the database (counter: often it is not)
2. Constraint may be not be implement at the database level (counter: then your database is wrong)
3. Some constraints cannot be checked by the database and will require application level validation anyway (good argument, but that's when it becomes business logic)
4. Easy to forget to do the error handling (also good argument, but... let's just *not forget*)

Reply:
Yeah your understanding is correct. 

Sometimes the entirety of the validation is just a data integrity thing. In the accounts case, there is no reasonable interpretation of multiple entries in players table for the same Discord user; there's nothing business logic here at all, it's purely data integrity, so we don't need to handle it at application level at all, so we don't.

If we do have things that are business logic (one example for this game, enemies with 0 health should be deleted from database so that the hunt ends) then we do that kind of thing at application level. An enemy with 0 health can still be interpreted meaningfully, so no need to prevent it at data level, so we don't enforce that in the database. 

The only argument against I have heard is people say "don't rely on database for that stuff" but it's a weak argument. Some people just don't trust databases for some reason. 🤷

Reply:
> What exactly is the alternative, handling all the errors on the code level before it gets sent to the database?

Yeah the alternative is to avoid seeing a database error at all costs. For accounts that would include:
```
const { rows: [{ exists }] } = db.sql`SELECT EXISTS (SELECT 1 FROM players WHERE id = ${id})`;
if (exists) {
  throw new Error('You have an account!');
}
db.sql`INSERT INTO players ...`;
```

That's just... extra work for everyone

Reply:
No thanks haha. We're going with what you proposed 👍